### PR TITLE
CA-220001: buffer not NULL terminated (CID:57557)

### PR DESCRIPTION
**Pull request by geosharath**
size of id is 40B, when msg length is > 39 it would cause
id not to be NULL terminated, hence we need to handle this
corner case.

Signed-off-by: <Sharath Babu sharath.babu@citrix.com>

### CODE REVIEW

File: drivers/block-nbd.c [hunk 1]
```diff
@@ -138,7 +139,17 @@ tdnbd_stash_passed_fd(int fd, char *msg, void *data)
...
-			sizeof(passed_fds[free_index].id));
+                       sizeof(passed_fds[free_index].id));
+
+	/*
+	 * Since size of id is 40B, if msg length is > 39 it would cause id
+	 * to overflow and not be null terminated. Hence, we need to handle 
+	 * this corner case.
+	 */
+
+	if(unlikely(strlen(msg) >= 40 ))
```
Comment (germanop) [inline drivers/block-nbd.c]: Writing explicit values is very fragile: consider sizeof or something else

**Comment by germanop** [comment]
In other places we are taking a different approach when it comes to buffer overflows: instead of truncating the message we print errors and return.

Would you confirm that in this case truncating is fine?

**Comment by geosharath** [comment]
this is not actually message, its the tapdisk driver image name. The only place I saw this being used is to get fd against a name in tdnbd_retreive_passed_fd() function. Here, we use strncmp so functionality wise it would not cause any problem (only first n bytes are compared). Please let me know if you feel this might break else where. However, I can think of a case where it might now work if we have first n bytes in the name as name and then they are different, is this case possible in the given environment ?

**Comment by germanop** [comment]
If you think it is not going to break anything that's fine.

However, using 40 and 39 explicitly it's a little bit fragile

**Comment by chandrikas** [comment]
I think truncating the msg is fine but agree with Germano. Let's use sizeof instead of explicit values.

**Comment by MarkSymsCtx** [comment]
This breaks Storage motion, needs rework.

**Comment by freddy77** [comment]
A buffer is NUL terminated, not NULL.

**Comment by MarkSymsCtx** [comment]
Buffer is a fixed length is not required to be NULL terminated.

**Pull request closed without merging**